### ISSUE TITLE
The operation couldn’t be completed. Connection refused

### ISSUE THREAD

**Issue by yingtao-futurewonder** [cjam/react-native-spotify-remote#75]
Below is my code. I got accessToken but when I call SpotifyRemote.connect(session.accessToken), I got this error

> Possible Unhandled Promise Rejection (id: 2):
> Error: Connection attempt failed.
> Stream error.
> The operation couldn’t be completed. Connection refused

```
        SpotifyAuth.getSession().then((session) => {
        if (session != undefined && session.accessToken != undefined) {
            SpotifyRemote
            .connect(session.accessToken)
            .then(() => {
                SpotifyRemote.playUri(`spotify:track:${songId}`);
            })
            .catch(this.onError);
        } else {
            console.log("session failed.=============")
        }

```

**Comment by yingtao-futurewonder** [comment]
It seems there is something wrong on the server-side.

**Comment by yingtao-futurewonder** [comment]
I found if I play a song on my Spotify app, then call SpotifyRemote.connect(session.accessToken), it works.

**Comment by cjam** [comment]
hey @yingtao-futurewonder , if this is on iOS, it could be that iOS has backgrounded spotify and thus shutting down the message server that this library connects to.  Have you seen the [playUri](https://cjam.github.io/react-native-spotify-remote/interfaces/spotifyapiconfig.html#playuri) configuration option for authorizing/connecting?

**Comment by yingtao-futurewonder** [comment]
@cjam Thanks for your reply! It works now after I add a playUrl to ApiConfig!

I found another problem though. When I shut down the Spotify app while playing a Spotify song in my app, then I call SpotifyRemote.connect, it will fail without restarting the Spotify app. Even I start the app again manually,  SpotifyRemote.connect still fails. Until I play a song inside the Spotify app, SpotifyRemote.connect works again.

**Comment by cjam** [comment]
@yingtao-futurewonder glad that was able to help.  Did you see in the docs I sent you that if you pass in `playUri: ""` it will attempt to play from the last position / song?  That might be less jarring for your application.  As for the second piece.  iOS is pretty aggressive with how it backgrounds applications which makes the connection to spotify somewhat challenging to maintain.  For the time being, in my app I've been checking to see if a connection has been lost and if it has, re-running the auth->playUri flow to make sure that spotify is started and not backgrounded.

Hope that helps.

**Comment by yingtao-futurewonder** [comment]
@cjam Thank you so much! it does help!

**Comment by MrBuggySan** [comment]
@cjam Thank you

**Comment by MrBuggySan** [comment]
Has anyone been able to connect back to the iOS Spotify client (after it becomes inactive) without switching back to Spotify app again? I thought I'd just cache the token and connect to Spotify client if somehow we get disconnected but it seems like we'd have to open the Spotify app each time.

**Comment by ryansaam** [comment]
I'm also having this issue. I'm using a real iOS (14.0.1) device. This is my code:

```javascript
// Api Config object, replace with your own applications client id and urls
const spotifyConfig: ApiConfig = {
	clientID: "fbc2a7bca8bf4a85846fc9d966a2b2fb",
	redirectURL: "spotify-ios-quick-start://spotify-login-callback",
	// tokenRefreshURL: "SPOTIFY_TOKEN_REFRESH_URL",
  // tokenSwapURL: "SPOTIFY_TOKEN_SWAP_URL",
  playURI: "",
	scopes: [ApiScope.AppRemoteControlScope, ApiScope.UserFollowReadScope]
}

// Initialize the library and connect the Remote
// then play an epic song
async function playEpicSong(){
	try{
    const session = await SpotifyAuth.authorize(spotifyConfig)
    console.log(session)
		await SpotifyRemote.connect(session.accessToken);
		await SpotifyRemote.playUri("");
    await SpotifyRemote.seek(58000);
	}catch(err){
		console.error("Couldn't authorize with or connect to Spotify",err)
	}
}
```

and the error is the same as OP.

**Comment by cjam** [comment]
@ryansaam Out of curiosity.  Is spotify closed or open when you try to run your code?  In the past I've found that closing Spotify and allowing the auth flow to open it back up has worked.  Not that I think it's a good solution to this problem, just more thoughts in terms of troubleshooting.

**Comment by cjam** [comment]
> Has anyone been able to connect back to the iOS Spotify client (after it becomes inactive) without switching back to Spotify app again? I thought I'd just cache the token and connect to Spotify client if somehow we get disconnected but it seems like we'd have to open the Spotify app each time.

@MrBuggySan yea some of the issues being faced here are at the iOS level.  iOS aggressively backgrounds apps.  In this case, when spotify is paused, iOS will background it within about 15-30 s from the time that it is no longer using the background audio.  When iOS backgrounds spotify, it kills spotify's message server that their SDK is based off of.  

One **super hacky** idea that I had for my app was to listen to the player state change events and detect when spotify is paused.  When you know that spotify is paused, periodically (before iOS has the chance to background spotify) **Play** and then immediately **Pause**, ideally it would be so short that the user wouldn't even here it.  If this wasn't possible, another thought is to play silence https://open.spotify.com/artist/1fbo6ce8Pg6ZPWlpE3OkfQ to keep spotify from backgrounding.  The downside to the silent track is that it would disrupt the current playing context and it will also skew the users popular tracks likely eventually mucking up the suggestions that spotify gives them.

In the ideal scenario, Spotify would prevent itself from being backgrounded if there was an app connected via its remote SDK.  For performance / battery life optimizations and App Store compliance they wouldn't want to do this all of the time but it seems reasonable that they could keep their app alive by some means if a client is connected.  

It sadly feels like the Spotify iOS sdk is not receiving updates which is sad, I'd love to make this library more robust but without some support from their end it's pretty tough.

Anyways, I digress.

**Comment by ryansaam** [comment]
> @ryansaam Out of curiosity. Is spotify closed or open when you try to run your code? In the past I've found that closing Spotify and allowing the auth flow to open it back up has worked. Not that I think it's a good solution to this problem, just more thoughts in terms of troubleshooting.

Yeah I tried a combination of that and it didn't work. So to answer your question I tried both.

You think it could be something to do with this `redirectURL: "spotify-ios-quick-start://spotify-login-callback"` provided by Spotify?

**Comment by cjam** [comment]
Hey @ryansaam, ah yes that would likely be it.  Please look at the docs here:  for setting up your iOS app. You basically need to register a callback URI with your app in the spotify dev dashboard. Then you need to add that URL to the `info.plist` under schemes so that iOS will know to open your app when it gets that URL scheme.

**Comment by ryansaam** [comment]
@cjam You didn't include the link for the docs. I assume you were talking about the docs for this library?

**Comment by cjam** [comment]
My bad. Here's the docs you'll need https://developer.spotify.com/documentation/ios/quick-start/

**Comment by ryansaam** [comment]
I got this to work. Just to clarify I had already Spotify's redirect url, spotify-ios-quick-start://spotify-login-callback, already white-listed in my developer console. What solved it for me was these parts of the Spotify iOS [Docs](https://developer.spotify.com/documentation/ios/quick-start/)

> Configure Info.plist
> Set -ObjC Linker Flag
> Add Bridging Header

After following them I was able to use react-native-spotify-remote library to play a song as in the [setup guide](https://github.com/cjam/react-native-spotify-remote#in-code) for this library. It works regardless of the Spotify app being opened or closed.

**Comment by cjam** [comment]
yea, they're are definitely some issues there that I've noticed, but I haven't been able to pin down if the cause lives in this library or Spotify's

**Comment by mguay22** [comment]
I ended up going with your first suggestion @cjam - looks something like this:

```
async pause() {
    await SpotifyRemote.pause();
    if (this.pauseInterval) {
      clearInterval(this.pauseInterval);
    }
    this.pauseInterval = setInterval(async () => {
      await SpotifyRemote.resume();
      await SpotifyRemote.pause();
    }, 15000);
  }
```

**Comment by sikky99** [comment]
> I ended up going with your first suggestion @cjam - looks something like this:
> 
> ```
> async pause() {
>     await SpotifyRemote.pause();
>     if (this.pauseInterval) {
>       clearInterval(this.pauseInterval);
>     }
>     this.pauseInterval = setInterval(async () => {
>       await SpotifyRemote.resume();
>       await SpotifyRemote.pause();
>     }, 15000);
> }
> 
> async play(songUri: string) {
>     if (this.pauseInterval) {
>       clearInterval(this.pauseInterval);
>     }
>     await SpotifyRemote.playUri(songUri);
> }
> ```
> 
> Works like a charm...
This works but do you have any other solution ?.
